jdbc-durable-state-store for postgres schemaName is ignored / leads to errors

Given:

- Pekko Persistence schema is created in postgres with schema changed (f.e. pekko)
- Application is configured for durable state actors

When:

Changing schemaName in .conf file for jdbc-durable-state-store to f.e. "pekko" leads to error:

```
org.apache.pekko.persistence.typed.state.internal.DurableStateStoreException: Failed to persist state with sequence number [1] for persistenceId [X]
...
Caused by: org.postgresql.util.PSQLException: ERROR: relation "durable_state" does not exist
```

--- a/core/src/main/scala/org/apache/pekko/persistence/jdbc/config/PekkoPersistenceConfig.scala
+++ b/core/src/main/scala/org/apache/pekko/persistence/jdbc/config/PekkoPersistenceConfig.scala
@@ -221,7 +221,13 @@ class DurableStateTableConfiguration(config: Config) {
   val batchSize: Int = config.getInt("batchSize")
   val schemaName: Option[String] = cfg.asStringOption("schemaName")
   val columnNames: DurableStateTableColumnNames = new DurableStateTableColumnNames(config)
-  val stateSequenceConfig = DurableStateSequenceRetrievalConfig(config)
+  val stateSequenceConfig: DurableStateSequenceRetrievalConfig = DurableStateSequenceRetrievalConfig(config)
+  private val schemaPrefix: String = schemaName.fold("")(schema => s"$schema.")
+
+  /**
+   * @since "1.1.1"
+   */
+  def schemaAndTableName: String = s"$schemaPrefix$tableName"
   override def toString: String = s"DurableStateTableConfiguration($tableName,$schemaName,$columnNames)"
 }
 

--- a/core/src/main/scala/org/apache/pekko/persistence/jdbc/state/DurableStateQueries.scala
+++ b/core/src/main/scala/org/apache/pekko/persistence/jdbc/state/DurableStateQueries.scala
@@ -48,7 +48,7 @@ import slick.jdbc.{ H2Profile, JdbcProfile, OracleProfile, PostgresProfile, SQLS
     durableStateTable.filter(_.persistenceId === persistenceId)
 
   private[jdbc] def insertDbWithDurableState(row: DurableStateTables.DurableStateRow, seqNextValue: String) = {
-    sqlu"""INSERT INTO #${durableStateTableCfg.tableName}
+    sqlu"""INSERT INTO #${durableStateTableCfg.schemaAndTableName}
             (
              #${durableStateTableCfg.columnNames.persistenceId},
              #${durableStateTableCfg.columnNames.globalOffset},
@@ -74,7 +74,7 @@ import slick.jdbc.{ H2Profile, JdbcProfile, OracleProfile, PostgresProfile, SQLS
   }
 
   private[jdbc] def updateDbWithDurableState(row: DurableStateTables.DurableStateRow, seqNextValue: String) = {
-    sqlu"""UPDATE #${durableStateTableCfg.tableName}
+    sqlu"""UPDATE #${durableStateTableCfg.schemaAndTableName}
            SET #${durableStateTableCfg.columnNames.globalOffset} = #$seqNextValue,
                #${durableStateTableCfg.columnNames.revision} = ${row.revision},
                #${durableStateTableCfg.columnNames.statePayload} = ${row.statePayload},

--- a/core/src/main/scala/org/apache/pekko/persistence/jdbc/state/SequenceNextValUpdater.scala
+++ b/core/src/main/scala/org/apache/pekko/persistence/jdbc/state/SequenceNextValUpdater.scala
@@ -57,7 +57,7 @@ import slick.sql.SqlStreamingAction
   import profile.api._
 
   def getSequenceNextValueExpr() =
-    sql"""SELECT nextval(pg_get_serial_sequence('#${durableStateTableCfg.tableName}', '#${durableStateTableCfg
+    sql"""SELECT nextval(pg_get_serial_sequence('#${durableStateTableCfg.schemaAndTableName}', '#${durableStateTableCfg
         .columnNames.globalOffset}'))""".as[
       String]
 }

--- a/core/src/test/scala/org/apache/pekko/persistence/jdbc/state/scaladsl/DurableStateStorePluginSpec.scala
+++ b/core/src/test/scala/org/apache/pekko/persistence/jdbc/state/scaladsl/DurableStateStorePluginSpec.scala
@@ -12,22 +12,12 @@ package org.apache.pekko.persistence.jdbc.state.scaladsl
 import com.typesafe.config.{ Config, ConfigFactory }
 import org.apache.pekko
 import pekko.actor._
-import pekko.persistence.jdbc.config.SlickConfiguration
-import pekko.persistence.jdbc.db.SlickDatabase
-import pekko.persistence.jdbc.testkit.internal.Postgres
-import pekko.persistence.jdbc.util.DropCreate
 import pekko.persistence.state.DurableStateStoreRegistry
-import pekko.util.Timeout
 import org.scalatest.matchers.should.Matchers
 import org.scalatest.wordspec.AnyWordSpecLike
-import org.scalatest.{ BeforeAndAfterAll, BeforeAndAfterEach }
-import org.scalatest.concurrent.Eventually.eventually
+import org.scalatest.BeforeAndAfterAll
 import org.scalatest.concurrent.ScalaFutures
-import org.scalatest.time.{ Millis, Seconds, Span }
-
 import slick.jdbc.{ H2Profile, JdbcProfile }
-
-import scala.concurrent.duration.DurationInt
 
 abstract class DurableStateStorePluginSpec(config: Config, profile: JdbcProfile)
     extends AnyWordSpecLike
@@ -52,90 +42,6 @@ abstract class DurableStateStorePluginSpec(config: Config, profile: JdbcProfile)
 
   override def afterAll(): Unit = {
     system.terminate().futureValue
-  }
-}
-
-abstract class DurableStateStorePostgresSchemaPluginSpec(val config: Config, profile: JdbcProfile)
-    extends AnyWordSpecLike
-    with BeforeAndAfterAll
-    with BeforeAndAfterEach
-    with Matchers
-    with ScalaFutures
-    with DropCreate
-    with DataGenerationHelper {
-
-  val schemaName: String = "pekko"
-  implicit val timeout: Timeout = Timeout(1.minute)
-  implicit val defaultPatience: PatienceConfig =
-    PatienceConfig(timeout = Span(60, Seconds), interval = Span(100, Millis))
-
-  val customConfig: Config = ConfigFactory.parseString(s"""
-    jdbc-durable-state-store {
-      tables {
-        durable_state {
-          schemaName = "pekko"
-        }
-      }
-    }
-  """)
-
-  implicit lazy val system: ExtendedActorSystem =
-    ActorSystem(
-      "test",
-      customConfig.withFallback(config)
-    ).asInstanceOf[ExtendedActorSystem]
-
-  lazy val db = SlickDatabase.database(
-    config,
-    new SlickConfiguration(config.getConfig("slick")), "slick.db"
-  )
-
-  private val createSchema = s"CREATE SCHEMA IF NOT EXISTS $schemaName;"
-  private val moveDurableStateTableToSchema = s"alter table public.durable_state set schema $schemaName;"
-  private val moveDurableStateTableToPublic = s"alter table $schemaName.durable_state set schema public;"
-  private val createSchemaAndMoveTable = s"${createSchema}${moveDurableStateTableToSchema}"
-
-  override def beforeAll(): Unit = {
-    dropAndCreate(Postgres)
-  }
-
-  override def beforeEach(): Unit = {
-    withStatement(_.execute(createSchemaAndMoveTable))
-  }
-
-  "A durable state store plugin" must {
-    "instantiate a JdbcDurableDataStore successfully" in {
-
-      val store = DurableStateStoreRegistry
-        .get(system)
-        .durableStateStoreFor[JdbcDurableStateStore[String]](JdbcDurableStateStore.Identifier)
-
-      store shouldBe a[JdbcDurableStateStore[_]]
-      store.system.settings.config shouldBe system.settings.config
-      store.profile shouldBe profile
-    }
-
-    "persist states successfully" in {
-
-      val store = DurableStateStoreRegistry
-        .get(system)
-        .durableStateStoreFor[JdbcDurableStateStore[String]](JdbcDurableStateStore.Identifier)
-
-      upsertManyForOnePersistenceId(store, "durable_state", "durable-t1", 1, 400).size shouldBe 400
-
-      eventually {
-        store.maxStateStoreOffset().futureValue shouldBe 400
-      }
-    }
-  }
-
-  override def afterEach(): Unit = {
-    withStatement(_.execute(moveDurableStateTableToPublic))
-  }
-
-  override def afterAll(): Unit = {
-    system.terminate().futureValue
-
   }
 }
 

--- a/core/src/test/scala/org/apache/pekko/persistence/jdbc/state/scaladsl/DurableStateStorePluginSpec.scala
+++ b/core/src/test/scala/org/apache/pekko/persistence/jdbc/state/scaladsl/DurableStateStorePluginSpec.scala
@@ -11,19 +11,20 @@ package org.apache.pekko.persistence.jdbc.state.scaladsl
 
 import com.typesafe.config.{ Config, ConfigFactory }
 import org.apache.pekko
-import org.apache.pekko.persistence.jdbc.config.SlickConfiguration
-import org.apache.pekko.persistence.jdbc.db.SlickDatabase
-import org.apache.pekko.persistence.jdbc.testkit.internal.Postgres
-import org.apache.pekko.persistence.jdbc.util.DropCreate
-import org.apache.pekko.util.Timeout
 import pekko.actor._
+import pekko.persistence.jdbc.config.SlickConfiguration
+import pekko.persistence.jdbc.db.SlickDatabase
+import pekko.persistence.jdbc.testkit.internal.Postgres
+import pekko.persistence.jdbc.util.DropCreate
+import pekko.persistence.state.DurableStateStoreRegistry
+import pekko.util.Timeout
 import org.scalatest.matchers.should.Matchers
 import org.scalatest.wordspec.AnyWordSpecLike
 import org.scalatest.{ BeforeAndAfterAll, BeforeAndAfterEach }
 import org.scalatest.concurrent.Eventually.eventually
 import org.scalatest.concurrent.ScalaFutures
 import org.scalatest.time.{ Millis, Seconds, Span }
-import pekko.persistence.state.DurableStateStoreRegistry
+
 import slick.jdbc.{ H2Profile, JdbcProfile }
 
 import scala.concurrent.duration.DurationInt
@@ -93,7 +94,6 @@ abstract class DurableStateStorePostgresSchemaPluginSpec(val config: Config, pro
   private val moveDurableStateTableToSchema = s"alter table public.durable_state set schema $schemaName;"
   private val moveDurableStateTableToPublic = s"alter table $schemaName.durable_state set schema public;"
   private val createSchemaAndMoveTable = s"${createSchema}${moveDurableStateTableToSchema}"
-  private val removeTableFromSchema = s"${moveDurableStateTableToPublic}"
 
   override def beforeAll(): Unit = {
     dropAndCreate(Postgres)
@@ -130,7 +130,7 @@ abstract class DurableStateStorePostgresSchemaPluginSpec(val config: Config, pro
   }
 
   override def afterEach(): Unit = {
-    withStatement(_.execute(removeTableFromSchema))
+    withStatement(_.execute(moveDurableStateTableToPublic))
   }
 
   override def afterAll(): Unit = {

--- a/core/src/test/scala/org/apache/pekko/persistence/jdbc/state/scaladsl/DurableStateStorePluginSpec.scala
+++ b/core/src/test/scala/org/apache/pekko/persistence/jdbc/state/scaladsl/DurableStateStorePluginSpec.scala
@@ -11,13 +11,22 @@ package org.apache.pekko.persistence.jdbc.state.scaladsl
 
 import com.typesafe.config.{ Config, ConfigFactory }
 import org.apache.pekko
+import org.apache.pekko.persistence.jdbc.config.SlickConfiguration
+import org.apache.pekko.persistence.jdbc.db.SlickDatabase
+import org.apache.pekko.persistence.jdbc.testkit.internal.Postgres
+import org.apache.pekko.persistence.jdbc.util.DropCreate
+import org.apache.pekko.util.Timeout
 import pekko.actor._
 import org.scalatest.matchers.should.Matchers
 import org.scalatest.wordspec.AnyWordSpecLike
-import org.scalatest.BeforeAndAfterAll
+import org.scalatest.{ BeforeAndAfterAll, BeforeAndAfterEach }
+import org.scalatest.concurrent.Eventually.eventually
 import org.scalatest.concurrent.ScalaFutures
+import org.scalatest.time.{ Millis, Seconds, Span }
 import pekko.persistence.state.DurableStateStoreRegistry
 import slick.jdbc.{ H2Profile, JdbcProfile }
+
+import scala.concurrent.duration.DurationInt
 
 abstract class DurableStateStorePluginSpec(config: Config, profile: JdbcProfile)
     extends AnyWordSpecLike
@@ -42,6 +51,91 @@ abstract class DurableStateStorePluginSpec(config: Config, profile: JdbcProfile)
 
   override def afterAll(): Unit = {
     system.terminate().futureValue
+  }
+}
+
+abstract class DurableStateStorePostgresSchemaPluginSpec(val config: Config, profile: JdbcProfile)
+    extends AnyWordSpecLike
+    with BeforeAndAfterAll
+    with BeforeAndAfterEach
+    with Matchers
+    with ScalaFutures
+    with DropCreate
+    with DataGenerationHelper {
+
+  val schemaName: String = "pekko"
+  implicit val timeout: Timeout = Timeout(1.minute)
+  implicit val defaultPatience: PatienceConfig =
+    PatienceConfig(timeout = Span(60, Seconds), interval = Span(100, Millis))
+
+  val customConfig: Config = ConfigFactory.parseString(s"""
+    jdbc-durable-state-store {
+      tables {
+        durable_state {
+          schemaName = "pekko"
+        }
+      }
+    }
+  """)
+
+  implicit lazy val system: ExtendedActorSystem =
+    ActorSystem(
+      "test",
+      customConfig.withFallback(config)
+    ).asInstanceOf[ExtendedActorSystem]
+
+  lazy val db = SlickDatabase.database(
+    config,
+    new SlickConfiguration(config.getConfig("slick")), "slick.db"
+  )
+
+  private val createSchema = s"CREATE SCHEMA IF NOT EXISTS $schemaName;"
+  private val moveDurableStateTableToSchema = s"alter table public.durable_state set schema $schemaName;"
+  private val moveDurableStateTableToPublic = s"alter table $schemaName.durable_state set schema public;"
+  private val createSchemaAndMoveTable = s"${createSchema}${moveDurableStateTableToSchema}"
+  private val removeTableFromSchema = s"${moveDurableStateTableToPublic}"
+
+  override def beforeAll(): Unit = {
+    dropAndCreate(Postgres)
+  }
+
+  override def beforeEach(): Unit = {
+    withStatement(_.execute(createSchemaAndMoveTable))
+  }
+
+  "A durable state store plugin" must {
+    "instantiate a JdbcDurableDataStore successfully" in {
+
+      val store = DurableStateStoreRegistry
+        .get(system)
+        .durableStateStoreFor[JdbcDurableStateStore[String]](JdbcDurableStateStore.Identifier)
+
+      store shouldBe a[JdbcDurableStateStore[_]]
+      store.system.settings.config shouldBe system.settings.config
+      store.profile shouldBe profile
+    }
+
+    "persist states successfully" in {
+
+      val store = DurableStateStoreRegistry
+        .get(system)
+        .durableStateStoreFor[JdbcDurableStateStore[String]](JdbcDurableStateStore.Identifier)
+
+      upsertManyForOnePersistenceId(store, "durable_state", "durable-t1", 1, 400).size shouldBe 400
+
+      eventually {
+        store.maxStateStoreOffset().futureValue shouldBe 400
+      }
+    }
+  }
+
+  override def afterEach(): Unit = {
+    withStatement(_.execute(removeTableFromSchema))
+  }
+
+  override def afterAll(): Unit = {
+    system.terminate().futureValue
+
   }
 }
 

--- a/integration-test/src/test/scala/org/apache/pekko/persistence/jdbc/integration/PostgresDurableStateStorePluginSpec.scala
+++ b/integration-test/src/test/scala/org/apache/pekko/persistence/jdbc/integration/PostgresDurableStateStorePluginSpec.scala
@@ -9,16 +9,114 @@
 
 package org.apache.pekko.persistence.jdbc.integration
 
-import com.typesafe.config.ConfigFactory
-import slick.jdbc.PostgresProfile
-import org.apache.pekko.persistence.jdbc.state.scaladsl.{
+import com.typesafe.config.{ Config, ConfigFactory }
+import org.apache.pekko
+import pekko.util.Timeout
+import pekko.actor.{ ActorSystem, ExtendedActorSystem }
+import pekko.persistence.jdbc.config.SlickConfiguration
+import pekko.persistence.jdbc.db.SlickDatabase
+import pekko.persistence.jdbc.state.scaladsl.{
+  DataGenerationHelper,
   DurableStateStorePluginSpec,
-  DurableStateStorePostgresSchemaPluginSpec
+  JdbcDurableStateStore
 }
+import pekko.persistence.jdbc.testkit.internal.Postgres
+import pekko.persistence.jdbc.util.DropCreate
+import pekko.persistence.state.DurableStateStoreRegistry
+import slick.jdbc.PostgresProfile
+import org.scalatest.concurrent.Eventually.eventually
+import org.scalatest.concurrent.ScalaFutures
+import org.scalatest.matchers.should.Matchers
+import org.scalatest.time.{ Millis, Seconds, Span }
+import org.scalatest.wordspec.AnyWordSpecLike
+import org.scalatest.{ BeforeAndAfterAll, BeforeAndAfterEach }
+import scala.concurrent.duration.DurationInt
 
 class PostgresDurableStateStorePluginSpec
     extends DurableStateStorePluginSpec(ConfigFactory.load("postgres-shared-db-application.conf"), PostgresProfile) {}
 
 class PostgresDurableStateStorePluginSchemaSpec
-    extends DurableStateStorePostgresSchemaPluginSpec(ConfigFactory.load("postgres-application.conf"),
-      PostgresProfile) {}
+    extends AnyWordSpecLike
+    with BeforeAndAfterAll
+    with BeforeAndAfterEach
+    with Matchers
+    with ScalaFutures
+    with DropCreate
+    with DataGenerationHelper {
+
+  val profile = PostgresProfile
+  val config = ConfigFactory.load("postgres-application.conf")
+  val schemaName: String = "pekko"
+  implicit val timeout: Timeout = Timeout(1.minute)
+  implicit val defaultPatience: PatienceConfig =
+    PatienceConfig(timeout = Span(60, Seconds), interval = Span(100, Millis))
+
+  val customConfig: Config = ConfigFactory.parseString(s"""
+    jdbc-durable-state-store {
+      tables {
+        durable_state {
+          schemaName = "pekko"
+        }
+      }
+    }
+  """)
+
+  implicit lazy val system: ExtendedActorSystem =
+    ActorSystem(
+      "test",
+      customConfig.withFallback(config)
+    ).asInstanceOf[ExtendedActorSystem]
+
+  lazy val db = SlickDatabase.database(
+    config,
+    new SlickConfiguration(config.getConfig("slick")), "slick.db"
+  )
+
+  private val createSchema = s"CREATE SCHEMA IF NOT EXISTS $schemaName;"
+  private val moveDurableStateTableToSchema = s"alter table public.durable_state set schema $schemaName;"
+  private val moveDurableStateTableToPublic = s"alter table $schemaName.durable_state set schema public;"
+  private val createSchemaAndMoveTable = s"${createSchema}${moveDurableStateTableToSchema}"
+
+  override def beforeAll(): Unit = {
+    dropAndCreate(Postgres)
+  }
+
+  override def beforeEach(): Unit = {
+    withStatement(_.execute(createSchemaAndMoveTable))
+  }
+
+  "A durable state store plugin" must {
+    "instantiate a JdbcDurableDataStore successfully" in {
+
+      val store = DurableStateStoreRegistry
+        .get(system)
+        .durableStateStoreFor[JdbcDurableStateStore[String]](JdbcDurableStateStore.Identifier)
+
+      store shouldBe a[JdbcDurableStateStore[_]]
+      store.system.settings.config shouldBe system.settings.config
+      store.profile shouldBe profile
+    }
+
+    "persist states successfully" in {
+
+      val store = DurableStateStoreRegistry
+        .get(system)
+        .durableStateStoreFor[JdbcDurableStateStore[String]](JdbcDurableStateStore.Identifier)
+
+      upsertManyForOnePersistenceId(store, "durable_state", "durable-t1", 1, 400).size shouldBe 400
+
+      eventually {
+        store.maxStateStoreOffset().futureValue shouldBe 400
+      }
+    }
+  }
+
+  override def afterEach(): Unit = {
+    withStatement(_.execute(moveDurableStateTableToPublic))
+  }
+
+  override def afterAll(): Unit = {
+    system.terminate().futureValue
+
+  }
+}

--- a/integration-test/src/test/scala/org/apache/pekko/persistence/jdbc/integration/PostgresDurableStateStorePluginSpec.scala
+++ b/integration-test/src/test/scala/org/apache/pekko/persistence/jdbc/integration/PostgresDurableStateStorePluginSpec.scala
@@ -11,11 +11,14 @@ package org.apache.pekko.persistence.jdbc.integration
 
 import com.typesafe.config.ConfigFactory
 import slick.jdbc.PostgresProfile
-import org.apache.pekko.persistence.jdbc.state.scaladsl.{DurableStateStorePluginSpec, DurableStateStorePostgresSchemaPluginSpec}
+import org.apache.pekko.persistence.jdbc.state.scaladsl.{
+  DurableStateStorePluginSpec,
+  DurableStateStorePostgresSchemaPluginSpec
+}
 
 class PostgresDurableStateStorePluginSpec
     extends DurableStateStorePluginSpec(ConfigFactory.load("postgres-shared-db-application.conf"), PostgresProfile) {}
 
-
 class PostgresDurableStateStorePluginSchemaSpec
-    extends DurableStateStorePostgresSchemaPluginSpec(ConfigFactory.load("postgres-application.conf"), PostgresProfile) {}
+    extends DurableStateStorePostgresSchemaPluginSpec(ConfigFactory.load("postgres-application.conf"),
+      PostgresProfile) {}

--- a/integration-test/src/test/scala/org/apache/pekko/persistence/jdbc/integration/PostgresDurableStateStorePluginSpec.scala
+++ b/integration-test/src/test/scala/org/apache/pekko/persistence/jdbc/integration/PostgresDurableStateStorePluginSpec.scala
@@ -11,7 +11,11 @@ package org.apache.pekko.persistence.jdbc.integration
 
 import com.typesafe.config.ConfigFactory
 import slick.jdbc.PostgresProfile
-import org.apache.pekko.persistence.jdbc.state.scaladsl.DurableStateStorePluginSpec
+import org.apache.pekko.persistence.jdbc.state.scaladsl.{DurableStateStorePluginSpec, DurableStateStorePostgresSchemaPluginSpec}
 
 class PostgresDurableStateStorePluginSpec
     extends DurableStateStorePluginSpec(ConfigFactory.load("postgres-shared-db-application.conf"), PostgresProfile) {}
+
+
+class PostgresDurableStateStorePluginSchemaSpec
+    extends DurableStateStorePostgresSchemaPluginSpec(ConfigFactory.load("postgres-application.conf"), PostgresProfile) {}

